### PR TITLE
feat: add simple projected coordinates

### PR DESCRIPTION
**Pull request by aplavin**
almost a copy of the [implementation](https://gitlab.com/aplavin/SkyImages.jl/-/blob/master/src/coords.jl) in SkyImages.jl

would be useful to upstream to SkyCoords, if this is in scope of the package

### CODE REVIEW

File: src/projected.jl [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+project(::Val{origin}, c::AbstractSkyCoords) where {origin} = ProjectedCoordsS{origin}(project(origin, c).offset)
+
+function project(origin::AbstractSkyCoords, c::AbstractSkyCoords)
+    cc = convert(typeof(origin), c)
+    Δlon = _center_angle(lon(cc) - lon(origin))
+    offset = SVector(Δlon * cos(lat(origin)), lat(cc) - lat(origin))
+    ProjectedCoords(origin, offset)
+end
+
+_center_angle(x) = mod(x + π, 2π) - π
```
Comment (mileslucas) [inline src/projected.jl:51]: This only appears to be used in one place, is it necessary to define this method?

File: src/projected.jl [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+end
+
+ProjectedCoordsS{C}(offset) where {C} = ProjectedCoordsS{C, eltype(offset)}(offset)
+constructorof(::Type{<:ProjectedCoordsS{C}}) where {C} = ProjectedCoordsS{C} ∘ tuple
+
+parent_coords_type(::Type{ProjectedCoords}) = Any
+parent_coords_type(::Type{ProjectedCoordsS}) = Any
+parent_coords_type(::Type{<:ProjectedCoords{TC}}) where {TC} = TC
+parent_coords_type(::Type{<:ProjectedCoordsS{TC}}) where {TC <: AbstractSkyCoords} = TC
+parent_coords_type(::Type{<:ProjectedCoordsS{origin}}) where {origin} = typeof(origin)
```
Comment (mileslucas) [inline src/projected.jl:23]: I'm not quite following why we need this- not used locally and is quite noisy- I would prefer if the interface could be used in a simpler way like

```julia
proj = ProjectedCoord() #...
parent_type = typeof(parent(proj))
```

this is the "obvious" way I would expect this to work.

**Review by mileslucas — COMMENTED** [review]
Thanks for working on this @aplavin! I get a little wary of code like this that starts doing lots of `Type` gymnastics- it's really easy to miss a corner case when constructors and converts get written this way. I would prefer to see an exhaustive test suite that uses a for loop to check all types work with projected coords (for both projected types).

I'm curious about your thoughts of defining `Base.parent` and using that in addition to or instead of `SkyCoords.origin`? Also, should `SkyCoords.separation` and `SkyCoords.position_angle` be defined for the projected offsets (i.e., we can measure the direct distance of the offset from the origin, but do we want that?)?

File: src/projected.jl [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+""" Projected coordinates with origin and offset.
+Approximation for small FoV. """
+struct ProjectedCoords{TC<:AbstractSkyCoords,T} <: AbstractProjectedCoords
+    origin::TC
+    offset::SVector{2,T}
+end
+
+""" Projected coordinates with origin (stored in the type) and offset.
+Approximation for small FoV. """
+struct ProjectedCoordsS{C,T} <: AbstractProjectedCoords
```
[the remark this answers is not in the document]
Comment (mileslucas) [reply src/projected.jl:12]: Hmmm well I'm definitely not a huge fan of putting entire objects into types, I think that greatly increases the complexity of the code, both for users and developers. I would definitely prefer a version that uses a field for storing the origin.

> and this technique cuts the memory usage of coordinates array in half.

do you have a benchmark to share for that? I'm curious if using `Ref` is a reasonable option as well.

File: src/projected.jl (hunk 2 above)
Comment (mileslucas) [reply src/projected.jl:23]: I still think this is a bit of an unnatural piece of API. Why not

```julia
C = typeof(origin(proj))
```

or, for example

```julia
function (proj::ProjectedCoords{C}) where C
	# ...
end
```

**Comment by mileslucas** [comment]
> I definitely prefer to keep origin either way, because it's the more intuitive name in coordinates context.

Sounds good

> > Also, should SkyCoords.separation and SkyCoords.position_angle be defined for the projected offsets
> 
> Now, offsets are just a SVector{2}s. Don't think we should overload these funcs for them...

I mean like so:

```julia
using LinearAlgebra
separation(proj::ProjectedCoords) = norm(proj.offset)
```

**Comment by icweaver** [comment]
Hi all, this finally fell onto my radar. I'm planning to polish some things up in this repo, but wanted to see if we can get this merged first. Is there anything left here that folks wanted to see? cc: @aplavin @giordano

File: src/projected.jl [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+struct ProjectedCoords{TC,T} <: AbstractProjectedCoords
+    origin::TC
+    offset::SVector{2,T}
+end
+
+origin(c::ProjectedCoords) = c.origin
+lon(c::AbstractProjectedCoords) = lon(origin(c)) + c.offset[1] / cos(lat(origin(c)))
+lat(c::AbstractProjectedCoords) = lat(origin(c)) + c.offset[2]
+
+Base.convert(::Type{T}, c::T) where {T<:AbstractProjectedCoords} = c
```
[the remark this answers is not in the document]
Comment (giordano) [reply src/projected.jl:14]: I'm still highly confused by this. "The Base method is less specific" doesn't say much to me if you don't point me to the more specific method which would override the Base one (of course besides this one).

[the remark this answers is not in the document]
Comment (icweaver) [reply src/projected.jl:14]: I had to play around with this a little bit, but I think I see Sasha's point. If I comment out:

```julia
Base.convert(::Type{T}, c::T) where {T<:AbstractProjectedCoords} = c
```

then this more specific [SkyCoords.jl method](https://github.com/aplavin/SkyCoords.jl/blob/573af89f33a8947235342c11cf909f98f0f8ccec/src/projected.jl#L15-L20) gets called when doing `convert(ProjectedCoords, cp::ProjectedCoords)` instead of Base's fast path.

Similar story for:

```julia
Base.convert(::Type{T}, c::T) where {T<:AbstractSkyCoords} = c
```

Commenting it out would lead to this more specific SkyCoords.jl method being called:

https://github.com/JuliaAstro/SkyCoords.jl/blob/560172ed30f96250961de22b2aba8c5cc0d6b2ed/src/types.jl#L133-L137

when doing something like `convert(ICRSCoords, c::ICRSCoords)`, which would silently return the wrong result

[the remark this answers is not in the document]
Comment (giordano) [reply src/projected.jl:14]: > then this more specific [SkyCoords.jl method](https://github.com/aplavin/SkyCoords.jl/blob/573af89f33a8947235342c11cf909f98f0f8ccec/src/projected.jl#L15-L20) gets called

Thanks, that's the explanation I was looking for.

> when doing something like `convert(ICRSCoords, c::ICRSCoords)`, which would silently return the wrong result

This is tangential to this PR, but getting the wrong result sounds bad?  Shouldn't that just be slow (doing unnecessary conversion) rather than giving wrong answer?  What's the problem with the rotation matrix in that case?  Shouldn't it be the identity matrix?

[the remark this answers is not in the document]
Comment (icweaver) [reply src/projected.jl:14]: Taking a closer look, I guess "wrong" is subjective. The identity matrix is indeed returned. It seems the different numbers I was getting was just from SkyCoords doing the flip for us when we go past 90 degree declination, like for `ICRSCoords(1, 2)`. Working directly in radians as another example:

```julia
julia> c = ICRSCoords(π/4, π/2 + π / 8) # π/8 past the pole
ICRSCoords{Float64}(0.7853981633974483, 1.9634954084936207)

julia> convert(ICRSCoords, c)
┌ Info: rotmat
│   rotmat(T, S) =
│    LinearAlgebra.UniformScaling{Bool}
└    true*I
ICRSCoords{Float64}(3.926990816987242, 1.1780972450961724)

julia> π/4 + π, π/2 - π/8 # Same as flipping over the pole
(3.9269908169872414, 1.1780972450961724)
```

So while it's nice that this just works™, idk if we want to keep this behavior since it could silently hide potential mistakes from the user accidentally entering declinations past the documented domain. It might be better to just validate on this, and have the user explicitly apply an offset if they need. This looks to be what astropy does.

I've opened a PR for the validation step https://github.com/JuliaAstro/SkyCoords.jl/pull/100, which can then be applied here

[the remark this answers is not in the document]
Comment (giordano) [reply src/projected.jl:14]: To be clear, fast paths are fine, and the new methods here are OK, if indeed needed to override other slow path methods, but I don't think having wrong results in even the slow paths is good (but that's orthogonal to this PR)

[the remark this answers is not in the document]
Comment (icweaver) [reply src/projected.jl:14]: Right, didn't mean to imply otherwise. I think we're good to go then

**Comment by icweaver** [comment]
Hey folks, I'm working on a redesign upstream to hopefully simplify some of the coordinate construction and conversion plumbing (#105).

Here's a follow-up PR showing what it would look like when introducing a new coordinate system (FK4, #101 rebased on #105), and a separate branch showing the [corresponding changes for ProjectedCoords](https://github.com/JuliaAstro/SkyCoords.jl/pull/106/changes). I like how it avoids the need for some of the extra conversion paths brought up by Mosè (https://github.com/JuliaAstro/SkyCoords.jl/pull/43#discussion_r1057002271) now, but this is a pretty substantial redesign that could really use another pair of eyes. Any additional comments welcome!
